android/src/main/java/com/polidea/flutter_ble_lib/delegate/DiscoveryDelegate.java

```diff
@@ -0,0 +1,194 @@
+                        call.<String>argument(ArgumentKey.DEVICE_IDENTIFIER),
+                        call.<String>argument(ArgumentKey.SERVICE_UUID),
+                        result
+                );
+                return;
+            case MethodName.GET_SERVICES:
+                getServices(
+                        call.<String>argument(ArgumentKey.DEVICE_IDENTIFIER),
+                        result
+                );
+                return;
+            case MethodName.GET_CHARACTERISTICS_FOR_SERVICE:
+                getCharacteristicsForService(call.<Integer>argument(ArgumentKey.SERVICE_ID), result);
+                return;
+            default:
+                throw new IllegalArgumentException(call.method + " cannot be handled by this delegate");
+        }
+    }
+
+    private void discoverAllServicesAndCharacteristics(String deviceId, String transactionId, final MethodChannel.Result result) {
+        final SafeMainThreadResolver resolver = new SafeMainThreadResolver<>(
+                new OnSuccessCallback<Object>() {
+                    @Override
+                    public void onSuccess(Object data) {
+                        result.success(null);
+                    }
+                },
+                new OnErrorCallback() {
+                    @Override
+                    public void onError(BleError error) {
+                        result.error(String.valueOf(error.errorCode.code), error.reason, null);
+                    }
+                });
+
+        adapter.discoverAllServicesAndCharacteristicsForDevice(deviceId, transactionId,
+                new OnSuccessCallback<Device>() {
+                    @Override
+                    public void onSuccess(Device data) {
+                        resolver.onSuccess(null);
+                    }
+                }, new OnErrorCallback() {
+                    @Override
+                    public void onError(BleError error) {
+                        resolver.onError(error);
+                    }
+                });
+    }
+
+    private void getCharacteristics(String deviceId, final String serviceUuid, final MethodChannel.Result result) {
+        adapter.getServicesForDevice(deviceId,
+                new OnSuccessCallback<Service[]>() {
+                    @Override
+                    public void onSuccess(final Service[] data) {
+                        for (final Service service : data) {
+                            if (service.getUuid().equals(UUID.fromString(serviceUuid))) {
+                                adapter.getCharacteristicsForService(service.getId(),
+                                        new OnSuccessCallback<Characteristic[]>() {
+                                            @Override
+                                            public void onSuccess(Characteristic[] data) {
+                                                CharacteristicsResponse characteristicsResponse = new CharacteristicsResponse(data, service);
```

Comment:
I know it is not Kotlin but IMHO here we have too many indentations

Reply:
Tried my best to reduce the number!